src/osd/PG.cc

```diff
@@ -311,7 +311,14 @@ void PG::proc_master_log(
   // See doc/dev/osd_internals/last_epoch_started
   if (oinfo.last_epoch_started > info.last_epoch_started)
     info.last_epoch_started = oinfo.last_epoch_started;
-  info.history.merge(oinfo.history);
+  if (info.history.merge(oinfo.history)) {
+    if (info.history.last_epoch_clean >= info.history.same_interval_since) {
+      dout(20) << __func__ << "clearing past_intervals" << dendl;
+      past_intervals.clear();
+      dirty_big_info = true;
+    }
+  }
+
```

Comment:
the same_interval_since is the start of one past_interval,  and info.history.last_epoch_clean is the epoch when pg chage to active+clean(PG::mark_clean can update,and replica received primary msg may be update);
In normal condition, last_epoch_clean can be bigger than same_interval_since;